File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/support/Action.java
package net.nemerosa.ontrack.model.support;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;

import java.net.URI;

@Data
@AllArgsConstructor(access = AccessLevel.PROTECTED)
public class Action {

    private final String id;
    private final String name;
    private final ActionType type;
    private final String uri;

    public static Action of(String id, String name, String uri, Object... parameters) {
        return new Action(id, name, ActionType.LINK, String.format(uri, parameters));
    }

    public static Action form(String id, String name, URI formUri) {
        return new Action(id, name, ActionType.FORM, formUri.toString());
    }

    public Action withUri(String uri) {
        return Action.of(id, name, uri);
    }

}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/support/Action.kt
package net.nemerosa.ontrack.model.support

import java.net.URI

@Deprecated("In V5, use UserMenuItem.")
class Action(
    val id: String,
    val name: String,
    val type: ActionType,
    val uri: String,
    /**
     * Group this action must be put into.
     */
    val group: String? = null,
    /**
     * Is this action enabled?
     */
    val enabled: Boolean = true,
) {

    fun withUri(uri: String): Action = Action(id, name, type, uri, group, enabled)
    fun withGroup(group: String): Action = Action(id, name, type, uri, group, enabled)

    companion object {

        @JvmStatic
        fun of(id: String, name: String, uri: String, vararg parameters: Any): Action {
            return Action(id, name, ActionType.LINK, String.format(uri, *parameters))
        }

        @JvmStatic
        fun form(id: String, name: String, formUri: URI): Action {
            return Action(id, name, ActionType.FORM, formUri.toString())
        }
    }

}